pkg/opensearch/response_parser.go

```diff
@@ -74,19 +77,206 @@ func (rp *responseParser) getTimeSeries() (*backend.QueryDataResponse, error) {
 		queryRes := backend.DataResponse{
 			Frames: data.Frames{},
 		}
-		props := make(map[string]string)
-		err := rp.processBuckets(res.Aggregations, target, &queryRes, props, 0)
-		if err != nil {
-			return nil, err
+
+		switch {
+		case target.Metrics[0].Type == rawDataType || target.Metrics[0].Type == rawDocumentType:
+			queryRes = processRawDataResponse(res, timeField, queryRes)
+		default:
+			props := make(map[string]string)
+			err := rp.processBuckets(res.Aggregations, target, &queryRes, props, 0)
+			if err != nil {
+				return nil, err
+			}
+			rp.nameFields(&queryRes.Frames, target)
+			rp.trimDatapoints(&queryRes.Frames, target)
 		}
-		rp.nameFields(&queryRes.Frames, target)
-		rp.trimDatapoints(&queryRes.Frames, target)
 
 		result.Responses[target.RefID] = queryRes
 	}
 	return result, nil
 }
 
+func processRawDataResponse(res *es.SearchResponse, timeField string, queryRes backend.DataResponse) backend.DataResponse {
+	propNames := make(map[string]bool)
+	docs := make([]map[string]interface{}, len(res.Hits.Hits))
+
+	for hitIdx, hit := range res.Hits.Hits {
+		var flattenedSource map[string]interface{}
+		if hit["_source"] != nil {
+			flattenedSource = flatten(hit["_source"].(map[string]interface{}), 10)
+		}
+
+		flattenedSource["_id"] = hit["_id"]
+		flattenedSource["_type"] = hit["_type"]
+		flattenedSource["_index"] = hit["_index"]
+		flattenedSource[timeField] = getTimestamp(hit, flattenedSource, timeField)
+
+		for key := range flattenedSource {
+			propNames[key] = true
+		}
+
+		docs[hitIdx] = flattenedSource
+	}
+	fields := processDocsToDataFrameFields(docs, propNames, timeField)
+
+	frames := data.Frames{}
+	frame := data.NewFrame("", fields...)
+	frames = append(frames, frame)
+
+	queryRes.Frames = frames
+	return queryRes
+}
+
+func getTimestamp(hit, source map[string]interface{}, timeField string) *time.Time {
+	timeString, ok := lookForFieldsInHit(hit, timeField)
+	if !ok {
+		timeString, ok = lookForTimeFieldInSource(source, timeField)
+		if !ok {
+			return nil
+		}
+	}
+
+	timeValue, err := time.Parse(time.RFC3339Nano, timeString)
+	if err != nil {
+		return nil
+	}
+
+	return &timeValue
+}
+
+func lookForFieldsInHit(hit map[string]interface{}, timeField string) (string, bool) {
+	if hit["fields"] != nil {
+		if fieldsMap, ok := hit["fields"].(map[string]interface{}); ok {
+			timesArray, ok := fieldsMap[timeField].([]interface{})
+			if !ok {
+				return "", false
+			}
+			if len(timesArray) == 1 {
+				if timeString, ok := timesArray[0].(string); ok {
+					return timeString, true
+				}
+			}
+		}
+	}
+	return "", false
+}
+
+func lookForTimeFieldInSource(source map[string]interface{}, timeField string) (string, bool) {
+	if source[timeField] != nil {
+		if timeString, ok := source[timeField].(string); ok {
+			return timeString, true
+		}
+	}
+
+	return "", false
+}
+
+func flatten(target map[string]interface{}, maxDepth int) map[string]interface{} {
+	output := make(map[string]interface{})
```

Comment:
Is there a reason this is a recursive function in this version? I think iterative ones usually end up being easier to read.

Reply:
Did you mean in this version vs Elasticsearch? Theirs is recursive.

Reply:
For some reason I thought it was iterative. Thanks for updating it!

Reply:
heya @iwysiu I think the recursive version, when refactored a bit, is much clearer. Here's what it looks like: https://github.com/grafana/grafana/pull/70302/files#diff-e269aab87b3357557da16d1389830e37476d46b47205d5c4c9b1d1d1d1473dc4R1044

Let me know if you prefer it!

Reply:
It looks nice! I'm still going to be annoying about commenting the arbitrary consts, lol. (Also if you're using it in 2 places it should be a variable)

Reply:
I don't really have a preference at this point, we can do the recursive version to match elasticsearch if that makes more sense to you.